managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1814,6 +1876,9 @@ public void operationComplete() {
                             mdEntry.newPosition);
                 }
 
+                INPROGRESS_MARKDELETE_PERSIST_POSITION_UPDATER.compareAndSet(ManagedCursorImpl.this,
+                        mdEntry.newPosition, null);
+
```

Comment:
internalMarkDelete is not a sync method, so persistPositionToLedger may persist 2 then persist 1

Reply:
@congbobo184 do you mean to say that there's a problem that should be resolved in some way? What do you suggest?

Reply:
we should make sure only one `persistPositionToLedger ` op in the same time. we should lock update persistentMarkDeletePosition and update INPROGRESS_MARKDELETE_PERSIST_POSITION_UPDATER like : 
```
        try {
            lock.writeLock().lock();
            if((persistentMarkDeletePosition == null ||
                    mdEntry.newPosition.compareTo(persistentMarkDeletePosition) > 0) && PENDING_MARK_DELETED_SUBMITTED_COUNT_UPDATER.compareAndSet(this, null, mdEntry.newPosition)) {
                persistPositionToLedger(cursorLedger, mdEntry, new VoidCallback() {
                    public void operationComplete() {
                        try {
                            lock.writeLock().lock();
                            INPROGRESS_MARKDELETE_PERSIST_POSITION_UPDATER.compareAndSet(ManagedCursorImpl.this,
                                    mdEntry.newPosition, null);
                            persistentMarkDeletePosition = mdEntry.newPosition;
                        } finally {
                            lock.writeLock().unlock();
                        }
                    }
                }
            }
        } finally {
            lock.writeLock().unlock();
        }
```
of cause you only use a flag not use INPROGRESS_MARKDELETE_PERSIST_POSITION_UPDATER is also can implement

Reply:
@congbobo184 Locks could have surprising side effects, such as deadlocks (example #11965). 

Please explain the reason why `persistentMarkDeletePosition` exists and the consistency guarantees for the value.

I think that using a single threaded execution model would be a better way to ensure consistency where ordering matters if that's the goal.

The changes in this PR already ensure eventual consistency at some level, but there's no ordering of the persisting operations, so there are possible gaps. The question is more about the impact if `persistentMarkDeletePosition` would be off by one etc.

Reply:
can use OrderedExecutor, if we don't ensure the order of persistentMarkDeletePosition, the pendingAck log will delete more then it will not recover to a correct state

Reply:
The change to use a single threaded model can be done in the future. This PR fixes the immediate problems.